benchmarks/src/main/scala/org/http4s/servlet/ServletIoBenchmarks.scala

```diff
@@ -0,0 +1,171 @@
+package org.http4s.servlet
+
+import cats.effect.IO
+import cats.effect.std.Dispatcher
+import cats.effect.unsafe.implicits.global
+
+import org.openjdk.jmh.annotations._
+import org.http4s.servlet.NonBlockingServletIo
+
+import java.io.ByteArrayInputStream
+import java.util.concurrent.TimeUnit
+import javax.servlet.{ServletInputStream, ReadListener}
+import javax.servlet.http.HttpServletRequest
+import scala.util.Random
+
+/** To do comparative benchmarks between versions:
+  *
+  * benchmarks/run-benchmark AsyncBenchmark
+  *
+  * This will generate results in `benchmarks/results`.
+  *
+  * Or to run the benchmark from within sbt:
+  *
+  * Jmh / run -i 10 -wi 10 -f 2 -t 1 cats.effect.benchmarks.AsyncBenchmark
+  *
+  * Which means "10 iterations", "10 warm-up iterations", "2 forks", "1 thread". Please note that
+  * benchmarks should be usually executed at least in 10 iterations (as a rule of thumb), but
+  * more is better.
+  */
+@State(Scope.Thread)
+@BenchmarkMode(Array(Mode.Throughput))
+@OutputTimeUnit(TimeUnit.SECONDS)
+class ServletIoBenchmarks {
+
+  @Param(Array("100000"))
+  var size: Int = _
+
+  @Param(Array("1000"))
+  var iters: Int = _
+
+  def servletRequest: HttpServletRequest = new HttpServletRequestStub(
+    new TestServletInputStream(Random.nextBytes(size))
+  )
+
+  @Benchmark
+  def reader() = {
+    val req = servletRequest
+    val servletIo = NonBlockingServletIo[IO](4096)
+
+    def loop(i: Int): IO[Unit] =
+      if (i == iters) IO.unit else servletIo.reader(req).compile.drain >> loop(i + 1)
+
+    loop(0).unsafeRunSync()
+  }
+
+  @Benchmark
+  def requestBody() = {
+    val req = servletRequest
+    val servletIo = NonBlockingServletIo[IO](4096)
+
+    Dispatcher
+      .sequential[IO]
```

Comment:
I think for a fair fight, we'd need to allocate this per request.

Reply:
Yes fair point 👍